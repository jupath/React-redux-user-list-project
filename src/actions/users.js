
export const usersAreLoading = bool => ({
  type: 'USERS_ARE_LOADING',
  isLoading: bool,
});

export const setUsers = users => ({
  type: 'SET_USERS',
  users,
});

export const startSetUsers = url => async (dispatch) => {
  dispatch(usersAreLoading(true));

  try {
    const response = await fetch(url);
    const data = await response.json();
    const results = await data.results;
    const users = await results.map(user => ({
      username: user.login.username,
      firstName: user.name.first,
      lastName: user.name.last,
      email: user.email,
      address: `${user.location.street}, ${user.location.city}, ${user.location.state}, ${user.location.postcode}`,
      dob: user.dob,
      cell: user.cell,
      image: user.picture.large,
    }));
    await dispatch(usersAreLoading(false));
    await dispatch(setUsers(users));
  } catch (e) {
    console.error(e);
  }
};

export const updateUser = (username, updatedData) => ({
  type: 'UPDATE_USER',
  username,
  updatedData,
});
