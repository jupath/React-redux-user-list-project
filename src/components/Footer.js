import React from 'react';

const Footer = () => (
  <footer className="footer pb-4 text-center">
    <h5>View <a href="https://github.com/jupath/React-redux-user-list-project/" target="_blank" rel="noopener noreferrer" className="font-weight-bold">Source on GitHub</a></h5>
  </footer>
);

export default Footer;
