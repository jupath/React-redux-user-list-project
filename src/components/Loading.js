import React from 'react';
import loader from '../images/spin.svg';

const Loading = () => (
  <div className="loading">
    <img src={loader} />
  </div>
);

export default Loading;