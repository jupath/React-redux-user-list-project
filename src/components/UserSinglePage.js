import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { Button } from 'reactstrap';
import UserSinglePageField from './UserSinglePageField';

export class UserSinglePage extends Component {

  render() {

    const dob = moment(this.props.user.dob).format('Do MMMM, YYYY');

    return (
      <div className="container">
        <div className="px-lg-6 mt-6 mb-4">
          <div className="row user-single p-3">
            <div className="pr-3">
              <img src={this.props.user.image} />
            </div>
            <div className="user-single__details">
              <h2 className="text-capitalize font-weight-bold">{this.props.user.firstName} {this.props.user.lastName}</h2>
              <div className="user-single__details__field">
                <div className="font-weight-bold pr-2">Email:</div>
                <UserSinglePageField data={this.props.user.email} field={'email'} username={this.props.user.username} isEditable={true} />
              </div>
              <div className="user-single__details__field">
                <div className="font-weight-bold pr-2">Cell:</div>
                <UserSinglePageField data={this.props.user.cell} field={'cell'} username={this.props.user.username} isEditable={true}/>
              </div>
              <div className="user-single__details__field text-capitalize">
                <div className="font-weight-bold pr-2">Address:</div>
                <UserSinglePageField data={this.props.user.address} field={'address'} username={this.props.user.username} isEditable={true} />
              </div>
              <div className="user-single__details__field">
                <div className="font-weight-bold pr-2">Day of birth:</div>
                <UserSinglePageField data={dob} field={'dob'} username={this.props.user.username} isEditable={false} />
              </div>
              <Button color="success" className="mt-3" onClick={() => this.props.history.push('/')}>Back to dashboard</Button>
            </div>
          </div>
        </div>
      </div>
    );
  };
}

const mapStateToProps = (state, props) => ({
  user: state.users.find((user) => user.username === props.match.params.username)
});

export default connect(mapStateToProps)(UserSinglePage);