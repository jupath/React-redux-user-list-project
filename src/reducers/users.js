export const usersAreLoading = (state = false, action) => {
  switch (action.type) {
    case 'USERS_ARE_LOADING':
      return action.isLoading;
    default:
      return state;
  }
};

export const users = (state = [], action) => {
  switch (action.type) {
    case 'SET_USERS':
      return action.users;
    case 'UPDATE_USER':
      return state.map((user) => {
        if (user.username === action.username) {
          return {
            ...user,
            ...action.updatedData,
          };
        }
        return user;
      });
    default:
      return state;
  }
};
