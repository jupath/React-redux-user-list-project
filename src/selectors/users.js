
export default (users, { text, orderBy }) => {
  return users.filter(user => {
    const regex = new RegExp( text, 'gi' );
    return user.firstName.match( regex ) || user.lastName.match( regex );
  }).sort((a, b) => {
    if( orderBy === 'desc' ) {
      return a.lastName < b.lastName ? 1 : -1;
    } else if ( orderBy === 'asc' ) {
      return a.lastName < b.lastName ? -1 : 1;
    }
  });
}