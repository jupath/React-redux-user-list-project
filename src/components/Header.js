import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button, Input } from 'reactstrap';
import { setOrderBy, setText } from '../actions/filters';

export class Header extends Component {
  handleOrderByClick = (orderBy) => {
    this.props.setOrderBy(orderBy);
  }

  handleSetText = (event) => {
    const text = event.target.value;
    this.props.setText(text);
  }

  render() {
    return (
      <header className="header">
        <div className="container text-center py-4">
          <h1>User list built with React and Redux</h1>
          <p>This app uses async-await to fetch users from <a href="https://randomuser.me/" target="_blank">randomuser.me</a> API</p>
          <div className="px-lg-7 pb-4">
            <Input type="text" placeholder="Search by name..." className="header__searchform" onChange={this.handleSetText} />
          </div>
          <div>
            <span>Order list by name: </span>
            <Button id="ascButton" color="success" className="mr-1" disabled={this.props.orderBy === 'asc'} onClick={() => this.handleOrderByClick('asc')}>Asc</Button>
            <Button id="descButton" color="success" className="mr-1" disabled={this.props.orderBy === 'desc'} onClick={() => this.handleOrderByClick('desc')}>Desc</Button>
            <Button id="defaultButton" color="success" disabled={this.props.orderBy === undefined} onClick={() => this.handleOrderByClick()}>Default</Button>
          </div>
        </div>
      </header>
    );
  }
}

const mapStateToProps = state => ({
  orderBy: state.filters.orderBy,
});

const mapDispatchToProps = dispatch => ({
  setOrderBy: orderBy => dispatch(setOrderBy(orderBy)),
  setText: text => dispatch(setText(text)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
