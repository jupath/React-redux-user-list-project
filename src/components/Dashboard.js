import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import filterUsers from '../selectors/users';
import Header from './Header';
import User from './User';
import Loading from './Loading';

export class Dashboard extends Component {
  renderUsers = () => {
    if (this.props.isLoading) {
      return <Loading />;
    }
    const users = this.props.users.map((user) => {
      const firstName = user.firstName.replace(this.props.filterText, `<span class="hl">${this.props.filterText}</span>`);
      const lastName = user.lastName.replace(this.props.filterText, `<span class="hl">${this.props.filterText}</span>`);

      const userData = {
        firstName,
        lastName,
        image: user.image,
        username: user.username,
        cell: user.cell,
      };

      return <User key={user.username} userData={userData} />;
    });
    return users;
  }

  render() {
    return (
      <div>
        <Header />
        <div className="container">
          <div className="row users pt-5">
            { this.renderUsers() }
          </div>
        </div>
      </div>
    );
  }
}

Dashboard.propTypes = {
  isLoading: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  users: filterUsers(state.users, state.filters),
  filterText: state.filters.text,
  isLoading: state.usersAreLoading,
});

export default connect(mapStateToProps)(Dashboard);
