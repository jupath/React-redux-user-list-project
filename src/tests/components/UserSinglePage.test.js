import React from 'react';
import { shallow } from 'enzyme';
import { UserSinglePage } from '../../components/UserSinglePage';
import users from '../fixtures/users';

test('should render UserSinglePage correctly', () => {
  const wrapper = shallow(<UserSinglePage user={users[0]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should handle Back to dashbord button on click', () => {
  const history = { push: jest.fn() };
  const wrapper = shallow(<UserSinglePage user={users[0]} history={history} />);
  wrapper.find('Button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
});