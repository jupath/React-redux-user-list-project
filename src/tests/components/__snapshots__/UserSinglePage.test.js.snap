// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render UserSinglePage correctly 1`] = `
<div
  className="container"
>
  <div
    className="px-lg-6 mt-6 mb-4"
  >
    <div
      className="row user-single p-3"
    >
      <div
        className="pr-3"
      >
        <img
          src="https://randomuser.me/api/portraits/men/83.jpg"
        />
      </div>
      <div
        className="user-single__details"
      >
        <h2
          className="text-capitalize font-weight-bold"
        >
          romain
           
          hoogmoed
        </h2>
        <div
          className="user-single__details__field"
        >
          <div
            className="font-weight-bold pr-2"
          >
            Email:
          </div>
          <Connect(UserSinglePageField)
            data="romain.hoogmoed@example.com"
            field="email"
            isEditable={true}
            username="lazyduck408"
          />
        </div>
        <div
          className="user-single__details__field"
        >
          <div
            className="font-weight-bold pr-2"
          >
            Cell:
          </div>
          <Connect(UserSinglePageField)
            data="(065)-247-9303"
            field="cell"
            isEditable={true}
            username="lazyduck408"
          />
        </div>
        <div
          className="user-single__details__field text-capitalize"
        >
          <div
            className="font-weight-bold pr-2"
          >
            Address:
          </div>
          <Connect(UserSinglePageField)
            data="1861 jan pieterszoon coenstraat, maasdriel, zeeland, 69217"
            field="address"
            isEditable={true}
            username="lazyduck408"
          />
        </div>
        <div
          className="user-single__details__field"
        >
          <div
            className="font-weight-bold pr-2"
          >
            Day of birth:
          </div>
          <Connect(UserSinglePageField)
            data="14th July, 1983"
            field="dob"
            isEditable={false}
            username="lazyduck408"
          />
        </div>
        <Button
          className="mt-3"
          color="success"
          onClick={[Function]}
          tag="button"
        >
          Back to dashboard
        </Button>
      </div>
    </div>
  </div>
</div>
`;
